### PR TITLE
PERF: reduce statelocks

### PR DESCRIPTION
**Pull request by tasinco**
Limit the use of statelock mutex.

### CODE REVIEW

File: network/network.go [hunk 1]
```diff
@@ -1052,3 +1006,61 @@ func (n *network) disconnected(p *peer) {
 		n.router.Disconnected(p.id)
 	}
 }
+
+type PeerElement struct {
```
Comment (aaronbuchwald) [inline network/network.go:1010]: The peer struct already has the ID, so I'm not sure the PeerElement struct is necessary

File: network/network.go [hunk 2]
```diff
@@ -1052,3 +1006,61 @@ func (n *network) disconnected(p *peer) {
...
+	peer        *peer
+	validatorId ids.ShortID
+}
+
+func (n *network) getPeers(validatorIDs ids.ShortSet) []PeerElement {
+	peers := make([]PeerElement, 0, 2)
```
Comment (aaronbuchwald) [inline network/network.go:1016]: Could we pre-allocate memory for this list based on the length of validatorIDs?

File: network/network.go [hunk 3]
```diff
@@ -792,15 +754,14 @@ func (n *network) gossip() {
 			continue
 		}
 
-		n.stateLock.Lock()
-		if n.closed {
-			n.stateLock.Unlock()
+		allpeers := n.getAllPeers()
```
Comment (aaronbuchwald) [inline network/network.go:757]: Could this variable be changed to use camel case?

File: network/peer.go [hunk 4]
```diff
@@ -204,31 +204,31 @@ func (p *peer) WriteMessages() {
...
 	if p.closed {
 		p.net.log.Debug("dropping message to %s due to a closed connection", p.id)
 		return false
 	}
 
 	msgBytes := msg.Bytes()
-	newPendingBytes := p.net.pendingBytes + len(msgBytes)
-	newConnPendingBytes := p.pendingBytes + len(msgBytes)
+	msgBytesLen := int64(len(msgBytes))
+	newPendingBytes := atomic.LoadInt64(&p.net.pendingBytes) + msgBytesLen
```
Comment (StephenButtolph) [inline network/peer.go:221]: I think there is a logical race introduced here. Not sure if we really care, but we should document it if we are intentionally allowing for a race.

File: network/network.go [hunk 5]
```diff
@@ -132,8 +132,8 @@ type network struct {
 
 	b Builder
 
-	stateLock       sync.Mutex
-	pendingBytes    int
+	stateLock       sync.RWMutex
```
Comment (StephenButtolph) [inline network/network.go:135]: Can we document that this stateLock should never be held when grabbing a peer lock?

File: network/peer.go [hunk 6]
```diff
@@ -53,6 +53,8 @@ type peer struct {
 
 	// unix time of the last message sent and received respectively
 	lastSent, lastReceived int64
+
+	lock sync.Mutex
```
Comment (StephenButtolph) [inline network/peer.go:57]: Can we document that only one peer lock should only be held at any given time by a thread of execution?

File: network/network.go (hunk 1 above)
Comment (tasinco) [reply network/network.go:1050]: I believe I need to keep this struct to hold the sent bool.  So we can properly log based on stephen's comments..
I did however see that the validator id is the key, and I use peer.id now.  (thx)

**Pull request closed without merging**